File: src/mcp/handler.py
import json
import os
from typing import Dict, Any, Optional
from src.qlik.auth import QlikAuth
from src.mcp.tools import (
    QlikGetAppsTool,
    QlikGetAppSheetsTool,
    QlikGetSheetChartsTool,
    QlikGetChartDataTool
)

class MCPHandler:
    def __init__(self):
        self.qlik_auth = QlikAuth()
        self.tools = {
            "qlik_get_apps": QlikGetAppsTool(),
            "qlik_get_app_sheets": QlikGetAppSheetsTool(),
            "qlik_get_sheet_charts": QlikGetSheetChartsTool(),
            "qlik_get_chart_data": QlikGetChartDataTool()
        }
    
    async def handle_request(self, body: Dict[str, Any], api_key: Optional[str] = None) -> Dict[str, Any]:
        import logging
        logger = logging.getLogger(__name__)
        
        request_id = body.get("id")
        method = body.get("method")
        params = body.get("params", {})
        
        # Usar API key do parâmetro ou do ambiente
        qlik_api_key = api_key or self.qlik_auth.get_api_key()
        
        if method == "initialize":
            return {
                "jsonrpc": "2.0",
                "id": request_id,
                "result": {
                    "protocolVersion": "2024-11-05",
                    "capabilities": {
                        "tools": {}
                    },
                    "serverInfo": {
                        "name": "qlik-cloud-mcp-server",
                        "version": "2.0.0",
                        "description": "Qlik Cloud MCP Server - Read-only queries using API key"
                    }
                }
            }
        
        elif method == "tools/list":
            tools_list = []
            for tool_name, tool_instance in self.tools.items():
                tools_list.append(tool_instance.get_schema())
            
            return {
                "jsonrpc": "2.0",
                "id": request_id,
                "result": {
                    "tools": tools_list
                }
            }
        
        elif method == "tools/call":
            tool_name = params.get("name")
            arguments = params.get("arguments", {})
            
            if tool_name not in self.tools:
                return {
                    "jsonrpc": "2.0",
                    "id": request_id,
                    "error": {
                        "code": -32601,
                        "message": f"Tool '{tool_name}' not found"
                    }
                }
            
            # Verificar se API key está disponível
            if tool_name.startswith("qlik_") and not qlik_api_key:
                return {
                    "jsonrpc": "2.0",
                    "id": request_id,
                    "error": {
                        "code": -32000,
                        "message": "Missing Qlik Cloud API key. Please configure QLIK_CLOUD_API_KEY."
                    }
                }
            
            try:
                tool_instance = self.tools[tool_name]
                result = await tool_instance.execute(arguments, qlik_api_key)
                
                return {
                    "jsonrpc": "2.0",
                    "id": request_id,
                    "result": {
                        "content": [
                            {
                                "type": "text",
                                "text": json.dumps(result, indent=2) if isinstance(result, (dict, list)) else str(result)
                            }
                        ]
                    }
                }
            except Exception as e:
                logger.error(f"Error executing tool {tool_name}: {str(e)}")
                return {
                    "jsonrpc": "2.0",
                    "id": request_id,
                    "error": {
                        "code": -32603,
                        "message": f"Error executing tool: {str(e)}"
                    }
                }
        
        else:
            return {
                "jsonrpc": "2.0",
                "id": request_id,
                "error": {
                    "code": -32601,
                    "message": f"Method '{method}' not found"
                }
            }

File: server/src/mcp/handler.ts
import { QlikRestClient } from "../qlik/restClient.js";
import { QlikEngineClient } from "../qlik/engineClient.js";
import * as listApps from "./tools/listApps.js";
import * as listSheets from "./tools/listSheets.js";

export type McpRequest = {
  jsonrpc?: string;
  id?: string | number | null;
  method?: string;
  params?: { name?: string; arguments?: Record<string, unknown> };
};

export type McpResponse = {
  jsonrpc: "2.0";
  id: string | number | null;
  result?: unknown;
  error?: { code: number; message: string };
};

const TOOLS: Record<
  string,
  { schema: { name: string; description: string; inputSchema: object }; execute: (a: unknown, args: Record<string, unknown>) => Promise<unknown> }
> = {
  [listApps.name]: {
    schema: {
      name: listApps.name,
      description: listApps.description,
      inputSchema: listApps.inputSchema,
    },
    execute: listApps.execute as (a: unknown, args: Record<string, unknown>) => Promise<unknown>,
  },
  [listSheets.name]: {
    schema: {
      name: listSheets.name,
      description: listSheets.description,
      inputSchema: listSheets.inputSchema,
    },
    execute: listSheets.execute as (a: unknown, args: Record<string, unknown>) => Promise<unknown>,
  },
};

export function createHandler(
  baseUrl: string,
  getToken: () => string | null
) {
  const rest = new QlikRestClient(baseUrl, getToken);

  return {
    async handle(req: McpRequest, tokenOverride: string | null): Promise<McpResponse> {
      const id = req.id ?? null;
      const token = () => tokenOverride ?? getToken();

      if (!req.method) {
        return { jsonrpc: "2.0", id, error: { code: -32600, message: "Missing 'method'" } };
      }

      if (req.method === "initialize") {
        return {
          jsonrpc: "2.0",
          id,
          result: {
            protocolVersion: "2024-11-05",
            capabilities: { tools: {} },
            serverInfo: {
              name: "qlik-mcp-server",
              version: "1.0.0",
              description: "Qlik Cloud MCP – list apps and sheets (read-only)",
            },
          },
        };
      }

      if (req.method === "tools/list") {
        return {
          jsonrpc: "2.0",
          id,
          result: {
            tools: Object.values(TOOLS).map((t) => t.schema),
          },
        };
      }

      if (req.method === "tools/call") {
        const name = req.params?.name;
        const args = (req.params?.arguments ?? {}) as Record<string, unknown>;
        if (!name) {
          return { jsonrpc: "2.0", id, error: { code: -32602, message: "Missing params.name" } };
        }
        const tool = TOOLS[name];
        if (!tool) {
          return { jsonrpc: "2.0", id, error: { code: -32601, message: `Unknown tool: ${name}` } };
        }
        const t = token();
        if (!t?.trim()) {
          return {
            jsonrpc: "2.0",
            id,
            error: {
              code: -32000,
              message: "Missing Qlik token. Set QLIK_TOKEN or Authorization: Bearer <token>.",
            },
          };
        }

        try {
          let out: unknown;
          if (name === listApps.name) {
            out = await tool.execute(rest, args);
          } else {
            const engine = new QlikEngineClient(baseUrl, () => t);
            try {
              out = await tool.execute(engine, args);
            } finally {
              engine.close();
            }
          }
          const text = typeof out === "string" ? out : JSON.stringify(out, null, 2);
          return {
            jsonrpc: "2.0",
            id,
            result: {
              content: [{ type: "text" as const, text }],
            },
          };
        } catch (err) {
          const msg = err instanceof Error ? err.message : String(err);
          return {
            jsonrpc: "2.0",
            id,
            error: { code: -32603, message: msg },
          };
        }
      }

      return {
        jsonrpc: "2.0",
        id,
        error: { code: -32601, message: `Method not found: ${req.method}` },
      };
    },
  };
}
